### PR TITLE
added github analysis

### PR DESCRIPTION
**Pull request by KarimKhalil33**
## 📝 Description

> Please include a summary of the change and which issue is fixed.  
> Include relevant motivation and context.  
> List any dependencies that are required for this change.
Enhanced GitHub Project Analysis with Advanced Pattern Detection and Resume Generation

This PR significantly enhances the GitHub project analysis capabilities by implementing comprehensive commit analysis, advanced technical keyword extraction, and intelligent resume generation from GitHub commit histories. The enhancement provides detailed insights into development patterns, practices, and technical expertise without requiring external LLM dependencies.

Advanced Technical Keyword Extraction:
Implemented comprehensive regex patterns covering 12+ technical domains
Extracts keywords from commit messages and file paths with intelligent filtering
Covers development actions, architecture, testing, DevOps, frontend/backend technologies, databases, mobile, data science, security, protocols, version control, and project management
Returns top 15 most relevant technical terms

GitHub Development Pattern Analysis:
Analyzes commit message patterns with configurable thresholds
Detects development practices: Test-Driven Development (>10%), Code Refactoring (>15%), Documentation-Focused (>5%)
Identifies project evolution based on file extension analysis
Categorizes commits by type: feature, bugfix, refactor, testing, documentation

Enhanced Resume Generation:
Generates professional resume bullets from GitHub contribution metrics
Includes contribution overview, file change analysis, technical focus areas
Highlights development patterns, project evolution, and quality practices
Incorporates impactful commit message sampling for key contributions

Improved GitHub Metrics Aggregation:
Enhanced file type classification with intelligent test file detection
Fixed file categorization logic to properly detect test files in any path location
Comprehensive role inference from commit file patterns
Duration calculation and contribution statistics

Supporting Infrastructure:
Updated role inference for commit file analysis
Enhanced main analysis workflow with GitHub project support
Comprehensive error handling for edge cases and malformed data

**Closes:** #276 
---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [x] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.

- [x] Enhanced Architectural Pattern Analysis - Tests commit message pattern detection with development practice thresholds
- [x] Technical Keyword Extraction - Validates extraction of architectural terms from commit messages and file paths
- [x] GitHub Resume Generation - Tests professional resume bullet creation from GitHub metrics
- [x] Edge Cases & Robustness - Tests empty inputs, malformed data, and minimal commit handling
- [x] Full Integration Workflow - End-to-end testing of GitHub project analysis pipeline
- [x] GitHub Analysis (No LLM) - Tests standalone functionality without external dependencies
- [x] Metrics Aggregation - Validates GitHub commit statistics and file classification
- [x] Commit Pattern Detection - Tests detection of test-driven development, refactoring practices, and documentation focus
- [x]  File Type Classification - Tests improved logic for detecting test files in any directory structure
- [x] Technical Domain Recognition - Tests extraction of DevOps, frontend, backend, database, and mobile development keywords
- [x] Role Inference - Tests detection of frontend, backend, DevOps, and data science roles from file changes

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

</details>

### CODE REVIEW

File: app/utils/code_analysis/code_analysis_utils.py [hunk 1]
```diff
@@ -85,6 +154,246 @@ def _detect_frameworks(imports: List[str]) -> List[str]:
     
     return detected_frameworks
 
+def analyze_github_development_patterns(commits: List[Dict]) -> Dict:
+    """
+    Analyze development patterns from GitHub commit history.
+    """
+    patterns = {
```
Comment (abstractafua) [inline app/utils/code_analysis/code_analysis_utils.py]: This is a good baseline start for git pattern analysis. I'm concerned that the `code_practices` and project_evolution` calculations may not be conclusive, but we can see how this performs add then re-iterate if necessary.

File: app/utils/code_analysis/code_analysis_utils.py [hunk 2]
```diff
@@ -85,6 +154,246 @@ def _detect_frameworks(imports: List[str]) -> List[str]:
...
+    else:
+        summary.append(f"Delivered {total_commits} focused commits in a concentrated development effort")
+    
+    # File and code changes
+    files_added = metrics.get('files_added', 0)
+    files_modified = metrics.get('files_modified', 0)
+    code_files = metrics.get('code_files_changed', 0)
+    
+    if code_files > 0:
+        summary.append(f"Implemented changes across {code_files} code files, with {files_added} new files created and {files_modified} existing files enhanced")
```
Comment (abstractafua) [inline app/utils/code_analysis/code_analysis_utils.py]: If no code files were changed this would read as - "Implemented changes across 0 code files, with 6 new files created and 8 existing files enhanced."

**Review by PaintedW0lf — COMMENTED** [review]
Great work done Karim!! this is gonna help us with GitHub commit analysis capabilities to our project's code analysis system, similar to how the existing system analyzes local code files.

Some of my own take aways that i think we can utilize for non code as well could be - 
Extensive Pattern Matching: 14 regex categories covering development actions, technologies, architectures
Fallback Strategy: Works with or without LLM 
Role Inference: Detects developer roles from file types (frontend, backend, devops, etc.)

File: app/utils/code_analysis/code_analysis_utils.py [hunk 3]
```diff
@@ -4,6 +4,75 @@
...
+    r'\b(git|github|gitlab|bitbucket|commit|commits|branch|branches|merge|merges|pull|request|requests|pr|fork|clone|push|rebase|cherry-pick|tag|tags|release|releases|version|versioning)\b',
+    # Project Management & Methodologies
+    r'\b(agile|scrum|kanban|sprint|sprints|story|stories|epic|epics|bug|bugs|issue|issues|feature|features|requirement|requirements|specification|specifications|documentation|docs|readme|changelog|license)\b'
+]
+    
+    for message in all_messages:
+        for pattern in tech_patterns:
+            matches = re.findall(pattern, message)
+            tech_terms.update(matches)
+
```
Comment (PaintedW0lf) [inline app/utils/code_analysis/code_analysis_utils.py]: Great job here, however i was wondering that You have 14 massive regex patterns running on every commit message and each pattern has 20-50 alternatives. I was worried if running 14 patterns on every message is expensive. 

So i searched up some alternatives- and came across 3 options - (I will let you decide what you think is acceptatble here )
suggestion
    
 Option 1: Compile patterns once (module level)
_COMPILED_PATTERNS = [re.compile(p) for p in tech_patterns]

 Option 2: Use single combined pattern with named groups
TECH_PATTERN = re.compile(r'\b(implement|add|create|react|vue|docker|...)\b')

 Option 3: Use simple word tokenization + set lookup (fastest)
TECH_KEYWORDS = {'implement', 'add', 'react', 'docker', ...}
words = set(message.lower().split())
tech_terms.update(words & TECH_KEYWORDS)

File: app/utils/code_analysis/code_analysis_utils.py [hunk 4]
```diff
@@ -4,6 +4,75 @@
 from collections import Counter, defaultdict
 import re
 
+
+def extract_technical_keywords_from_github(commits: List[Dict]) -> List[str]:
```
Comment (PaintedW0lf) [inline app/utils/code_analysis/code_analysis_utils.py]: Good job on the function here i also noticed that 
extract_technical_keywords_from_github() and extract_technical_keywords_from_parsed() share similar logic like Both split camelCase, filter by length and  return top 15 keywords
what do you think of Extracting shared logic into helper functions?

File: app/utils/code_analysis/code_analysis_utils.py [hunk 5]
```diff
@@ -4,6 +4,75 @@
...
+    
+    # Extract meaningful file names
+    for filename in all_file_names:
+        if filename and len(filename) > 2 and not filename.endswith(('.git', '.md')):
+            # Remove extensions and split camelCase
+            name_without_ext = filename.split('.')[0]
+            words = re.findall(r'[A-Z][a-z]+|[a-z]+|[A-Z]+(?=[A-Z][a-z]|\b)', name_without_ext)
+            tech_terms.update(word.lower() for word in words if len(word) > 2)
+    
+    return sorted(list(tech_terms))[:15]
```
[the remark this answers is not in the document]
Comment (KarimKhalil33) [reply app/utils/code_analysis/code_analysis_utils.py]: Correct, we do that once we sort in order to get the top 15 tech terms we catch within a project

File: app/utils/code_analysis/code_analysis_utils.py (hunk 1 above)
Comment (KarimKhalil33) [reply app/utils/code_analysis/code_analysis_utils.py]: sounds good will keep a close eye to it once we start testing the overall analysis

File: app/utils/code_analysis/code_analysis_utils.py (hunk 2 above)
Comment (KarimKhalil33) [reply app/utils/code_analysis/code_analysis_utils.py]: added a fallback. Please let me know if it looks fine now

**Review by abstractafua — APPROVED** [review]
Changes look good on my end
